src/D2L.CodeStyle.Analyzers/Async/Generator/AsyncToSyncMethodTransformer.cs

```diff
@@ -0,0 +1,135 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+internal sealed class AsyncToSyncMethodTransformer : SyntaxTransformer {
+	public AsyncToSyncMethodTransformer( SemanticModel model ) : base( model ) { }
+
+	public TransformResult<MethodDeclarationSyntax> Transform( MethodDeclarationSyntax decl ) {
+		// TODO: remove CancellationToken parameters
+		decl = decl.WithAttributeLists( ReplaceGenerateSyncAttribute( decl.AttributeLists ) )
+			.WithModifiers( RemoveAsyncModifier( decl.Modifiers ) )
+			.WithIdentifier( RemoveAsyncSuffix( decl.Identifier ) )
+			.WithReturnType( TransformReturnType( decl.ReturnType ) )
+			.WithExpressionBody( Transform( decl.ExpressionBody ) )
+			.WithBody( Transform( decl.Body ) );
+
+		return GetResult( decl );
+	}
+
+	private SyntaxList<AttributeListSyntax> ReplaceGenerateSyncAttribute(
+		SyntaxList<AttributeListSyntax> attributeLists
+	) => TransformAll( attributeLists, ReplaceGenerateSyncAttribute );
+
+	private AttributeListSyntax? ReplaceGenerateSyncAttribute(
+		AttributeListSyntax attributeList
+	) => attributeList.WithAttributes(
+		TransformAll( attributeList.Attributes, ReplaceGenerateSyncAttribute )
+	);
+
+	private AttributeSyntax ReplaceGenerateSyncAttribute(
+		AttributeSyntax attribute
+	) {
+		// Attributes other than [GenerateSync] pass through
+		if( !SyncGenerator.IsGenerateSyncAttribute( m_model, attribute ) ) {
+			return attribute;
+		}
+
+		// Turn [GenerateSync] into [Blocking].
+		// This makes some assumptions about the ambient using directives.
+		return attribute.WithName(
+			SyntaxFactory.ParseName( "Blocking" ).WithTriviaFrom( attribute.Name )
+		);
+	}
+
+	private SyntaxTokenList RemoveAsyncModifier( SyntaxTokenList modifiers ) =>
+		TransformAll(
+			modifiers,
+			static token => token.IsKind( SyntaxKind.AsyncKeyword ) ? null : token
+		);
+
+	private SyntaxToken RemoveAsyncSuffix( SyntaxToken ident ) {
+		if( !ident.ValueText.EndsWith( "Async" ) || ident.ValueText == "Async" ) {
+			ReportDiagnostic(
+				Diagnostics.ExpectedAsyncSuffix,
+				ident.GetLocation(),
+				ident.ValueText
+			);
+
+			return ident;
+		}
+
+		const int AsyncSuffixLength = 5;
+
+		return SyntaxFactory.Identifier(
+			ident.ValueText.Substring( 0, ident.ValueText.Length - AsyncSuffixLength )
+		).WithTriviaFrom( ident );
+	}
+
+	private TypeSyntax TransformReturnType( TypeSyntax returnType ) {
+		var returnTypeInfo = m_model.GetTypeInfo( returnType );

```

Comment:
Another usage of `SemanticModel`. This one feels more important? If we only do it by string then we won't handle all the cases, but maybe that's fine. If you're using `[GenerateSync]` you should just be reasonable?

I'm on the fence about this. WDYT?

Reply:
I don't think it would be that crazy to require "Task" => "void" and "Task<...>".Match( /Task<(.+)>/)[1]. Maybe eventually add ValueTask as well. Any custom awaitable support would be insane for this to handle probably. Maybe just make the diagnostic a little more specific: "Expected return type to be exactly 'Task' or 'Task<...>' (no qualified names)."?

Reply:
Said simpler, I think I would okay with these being stringy as long as the diagnostic messaging is clear.

I'm also okay with it using the model right now, echoing the previous sentiment.

Reply:
Yep, I already have that diagnostic below and definitely wasn't intended to support anything more general than the main task types.